Data/Pool.hs

```diff
@@ -263,46 +295,56 @@ withResource pool act = control $ \runInIO -> mask $ \restore -> do
             destroyResource pool local resource
   putResource local resource
   return ret
-#if __GLASGOW_HASKELL__ >= 700
 {-# INLINABLE withResource #-}
-#endif
 
 -- | Take a resource from the pool, following the same results as
 -- 'withResource'. Note that this function should be used with caution, as
 -- improper exception handling can lead to leaked resources.
 --
+-- * If a timeout is set, then this function will only block for
+-- 'timeout' seconds before throwing a 'TimeoutException'.
+--
 -- This function returns both a resource and the @LocalPool@ it came from so
 -- that it may either be destroyed (via 'destroyResource') or returned to the
 -- pool (via 'putResource').
 takeResource :: Pool a -> IO (a, LocalPool a)
 takeResource pool@Pool{..} = do
+  timeoutSync <- newTVarIO False
+  mgr <- Event.getSystemTimerManager
+  timeoutKey <-
+    Event.registerTimeout mgr (toSeconds timeout)
+      . atomically
+      . writeTVar timeoutSync
+      $ True
   local@LocalPool{..} <- getLocalPool pool
   resource <- liftBase . join . atomically $ do
-    ents <- readTVar entries
-    case ents of
-      (Entry{..}:es) -> writeTVar entries es >> return (return entry)
-      [] -> do
-        used <- readTVar inUse
-        when (used == maxResources) retry
-        writeTVar inUse $! used + 1
-        return $
-          create `onException` atomically (modifyTVar_ inUse (subtract 1))
+    stop <- readTVar timeoutSync
```

Comment:
Curious why not use `System.Timeout`? Sorry if this was already discussed.

Reply:
My understanding is that the event loop is more efficient since `System.Timeout` uses an extra thread every time.